lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs

```diff
@@ -722,3 +729,41 @@ hex = convertToBase Base16
 -- | Decode a 'ByteString' from base16
 fromHex :: ByteArray bout => ByteString -> Either String bout
 fromHex = convertFromBase Base16
+
+-- | Extracts the raw xprv bytes as @private key <> chain code@, i.e. without
+-- the public key.
+unXPrvStripPub :: XPrv -> ByteString
+unXPrvStripPub = stripPub . unXPrv
+  where
+    -- Converts  prv <> pub <> cc
+    -- To        prv <>        cc
+    stripPub :: ByteString -> ByteString
+    stripPub xprv' = prv <> chainCode
+      where
+        (prv, rest) = BS.splitAt 64 xprv'
+        (_pub, chainCode) = BS.splitAt 32 rest
+
+data ErrXPrvFromStrippedPubXPrv
+    = ErrInputLengthMismatch Int Int -- ^ Expected, Actual
+    | ErrInternalError String
+    deriving (Eq, Show)
+
+-- Create a @XPrv@ from a prv <> cc bytestring.
+xPrvFromStrippedPubXPrv :: ByteString -> Either ErrXPrvFromStrippedPubXPrv XPrv
+xPrvFromStrippedPubXPrv x =
+    let
+        res = toXPrv $ CC.encryptedCreateDirectWithTweak x pass
```

Comment:
This won't work for Byron keys though. The "tweak" mention here does actually change a few bits that could have a disastrous effects on legacy keys:

```
	secret_key[0] &= 248;   /* clears the bottom 3 bits */
	secret_key[31] &= 0x1F; /* clears the 3 highest bits */
	secret_key[31] |= 64;   /* set the 2nd highest bit */
```

We should instead go for `wallet_encrypted_initialize` which simply computes the public key from an already tweaked private key. Yet, this is not exported via the FFI on Haskell. So, another option would be to make this function depends on the target version (V1 vs V2) and use the appropriate creation function.

Reply:
> This won't work for Byron keys though. 

What won't work? For _unencrypted_ keys, all the roundtrips are passing. The CLISpec roundtrips include Byron keys.

I think I see how we could throw in `unXPrvStripPub` if the keys are _encrypted_, like you suggested elsewhere.

Reply:
The issue has nothing to do with encryption here. Keys are a random sequence of bytes. However, some **bits** in the first and last **byte** are tweaked in Icarus key for some security reasons.

Looking back into the generation methods (https://gist.github.com/KtorZ/b086754a15718affcc1b5c03c12c078a#byron), the Byron approach isn't exactly the same but should work all-the-same **for a key that has already been generated**. Soooooo. Okay. Nevermind.

Reply:
> Byron approach isn't exactly the same but should work all-the-same for a key that has already been generated.

Are you saying that the difference between bit-tweaking of Byron and Icarus keys shouldn't matter in this case?

If this tweak "clears the 3 highest bits" but Byron keys only have the highest bit cleared… shouldn't that modify the key? It looks like the 2nd highest bit is set, but the 3rd highest bit should differ?

Although roundtrips for non-encrypted Byron-keys haven't failed so far. I would have expected them to fail in that case.

> for a key that has already been generated

Not sure I get this.

Reply:
I tested with some old mnemonics and they seem to work (have roundtripping root keys) though.